### PR TITLE
[DEV-1741] HCM interventions data model change

### PR DESCRIPTION
**Pull request by Esaotuelr**
# Description

Added the following messages/classes/types:
  * `StaticVarCompensator`
  * `SVCControlMode`
  * `EndDeviceFunction`
  * `PanDemandResponseFunction`
  * `ControlledAppliance`
  * `EndDeviceFunctionKind`
  * `AssetFunction`
  * `BatteryControl`

Added new attributes to `RegulatingControl`:
  * `ctPrimary`
  * `minTargetDeadband`

Added repeated attribute of `batteryControls` to `BatteryUnit`.

Added repeated attribute of `endDeviceFunctions` to `EndDevice`.

# Test Steps

Explain in detail how your reviewer can test the changes proposed in this PR. If it cannot be tested, leave an explanation on why.

# Checklist

If any of these are not applicable, strikethrough the line `~like this~`. **Do not delete it!**. Let the reviewer decide if you should have done it.

### Code
- [x] I have performed a self review of my own code (including checking issues raised when creating the PR).
- [x] I have added/updated unit tests for these changes, and if not I have explained why they are not necessary.
- [x] I have commented my code in any hard-to-understand or hacky areas.
- [x] I have handled all new warnings generated by the compiler or IDE.
- [x] I have rebased onto the target branch (usually main).

### Documentation
- [x] I have updated the changelog.
- [x] I have updated any documentation required for these changes.

# Breaking Changes
- [x] I have considered if this is a breaking change and will communicate it with other team members if so.

I think this shouldntt be a breaking change as we are only adding stuff.

### CODE REVIEW

**Comment by ryanjcoleman** [comment]
Task linked: [DEV-1741 Update evolve-grpc](https://app.clickup.com/t/6929263/DEV-1741)

File: proto/zepben/protobuf/cim/extensions/iec61968/metering/PanDemandResponseFunction.proto [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+import "zepben/protobuf/cim/iec61968/metering/EndDeviceFunctionKind.proto";
+import "zepben/protobuf/cim/iec61968/metering/ControlledAppliance.proto";
+
+/**
+ * [ZBEX] PAN function that an end device supports, distinguished by 'kind'.
+ */
+message PanDemandResponseFunction {
+
+    /**
+     * EndDeviceFunction for this PanDemandResponseFunction.
```
Comment (charlta) [inline proto/zepben/protobuf/cim/extensions/iec61968/metering/PanDemandResponseFunction.proto:26]: todo: Missing `fields` in comment makes this look like it owns one, not is one.
```suggestion
     * EndDeviceFunction fields for this PanDemandResponseFunction.
```

**Comment by charlta** [comment]
I updated the name of the PR and the description to be more useful than "data model change".

I didn't change the branch name (which I should have) as it will close the PR

**Comment by charlta** [comment]
Added ewb-core as reviewers as it shouldn't have been sent straight to me.

File: proto/zepben/protobuf/cim/extensions/iec61968/metering/PanDemandResponseFunction.proto [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+    /**
+     * [ZBEX] Kind of this function.
+     */
+    cim.iec61968.metering.EndDeviceFunctionKind kind = 2;
+
+    /**
+     * [ZBEX] ControlledAppliance is not an IdentifiedObject, so in terms of database storage can be collapsed into the pan_demand_response_functions table.
+     * It could also be a data class and "interned", as there's likely to only be a couple of these, and there's only a maximum of 4096 combinations (and practically much less than that).
+     */
```
Comment (charlta) [inline proto/zepben/protobuf/cim/extensions/iec61968/metering/PanDemandResponseFunction.proto:38]: todo: Incorrect description. You have placed the `notes` field here, not the `description`.

File: proto/zepben/protobuf/cim/extensions/iec61968/metering/PanDemandResponseFunction.proto [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ */
+
+
+syntax = "proto3";
+option java_multiple_files = true;
+option java_package = "com.zepben.protobuf.cim.extensions.iec61968.metering";
+option csharp_namespace = "Zepben.Protobuf.CIM.Extensions.Iec61968.Metering";
```
Comment (charlta) [inline proto/zepben/protobuf/cim/extensions/iec61968/metering/PanDemandResponseFunction.proto:13]: todo: incorrect casing in package name.
```suggestion
option csharp_namespace = "Zepben.Protobuf.CIM.Extensions.IEC61968.Metering";
```

**Comment by Esaotuelr** [comment]
> I updated the name of the PR and the description to be more useful than "data model change".
> 
> I didn't change the branch name (which I should have) as it will close the PR

Thanks for that. Also I wasn't sure who could review this so I did just straight put your name against it for review. I'll put ewb-core for the rest of the other changes.

**Review by charlta — CHANGES_REQUESTED** [review]
I have barely looked at the PR and found some very sloppy mistakes.

Given this and the lazy nature of the PR name and description I feel this requires a more thorough self review before I look at it any further.

File: proto/zepben/protobuf/cim/extensions/iec61968/metering/PanDemandResponseFunction.proto (hunk 2 above)
Comment (Esaotuelr) [reply proto/zepben/protobuf/cim/extensions/iec61968/metering/PanDemandResponseFunction.proto]: Fixed

File: proto/zepben/protobuf/cim/extensions/iec61968/metering/PanDemandResponseFunction.proto (hunk 1 above)
Comment (Esaotuelr) [reply proto/zepben/protobuf/cim/extensions/iec61968/metering/PanDemandResponseFunction.proto]: Fixed and others.

File: proto/zepben/protobuf/cim/extensions/iec61968/metering/PanDemandResponseFunction.proto (hunk 3 above)
Comment (Esaotuelr) [reply proto/zepben/protobuf/cim/extensions/iec61968/metering/PanDemandResponseFunction.proto]: Fixed and others.

File: spec/ewb/IEC61970/Base/Wires/StaticVarCompensator.yaml [hunk 4]
```diff
@@ -0,0 +1,27 @@
+name: StaticVarCompensator
```
Comment (charlta) [inline spec/ewb/IEC61970/Base/Wires/StaticVarCompensator.yaml:1]: todo: Again, not copied then modified from the base spec version. This is evident with the format of the description and the difference in units.

File: spec/ewb/extensions/IEC61968/Metering/PanDemandResponseFunction.yaml [hunk 5]
```diff
@@ -0,0 +1,11 @@
+name: PanDemandResponseFunction
+description: PAN function that an end device supports, distinguished by 'kind'.
```
Comment (charlta) [inline spec/ewb/extensions/IEC61968/Metering/PanDemandResponseFunction.yaml:2]: todo: Missing ZBEX in descriptions. Search the repo for other uses of this. This applies to other extensions.

File: spec/ewb/extensions/IEC61970/Base/Wires/BatteryControl.yaml [hunk 6]
```diff
@@ -0,0 +1,18 @@
+name: BatteryControl
+description: Describes behaviour specific to controlling batteries.
+attributes:
+  - name: chargingRate
+    type: Double
```
Comment (charlta) [inline spec/ewb/extensions/IEC61970/Base/Wires/BatteryControl.yaml:5]: todo: Should link to appropriate units (as per description). i.e. [PerCent](https://zepben.github.io/evolve/docs/cim/ewb/IEC61970/Base/Domain/PerCent) in this case. This also applies to other units available in the `IEC61970/Base/Domain` package.

**Review by charlta — CHANGES_REQUESTED** [review]

File: proto/zepben/protobuf/cim/extensions/iec61970/base/wires/BatteryControlMode.proto [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+ */
+
+syntax = "proto3";
+option java_multiple_files = true;
+option java_package = "com.zepben.protobuf.cim.extensions.iec61970.wires";
+option csharp_namespace = "Zepben.Protobuf.CIM.Extensions.IEC61970.Wires";
+package zepben.protobuf.cim.extensions.iec61970.wires;
+
+/**
+ * Kind of battery control.
```
[the remark this answers is not in the document]
Comment (Esaotuelr) [reply proto/zepben/protobuf/cim/extensions/iec61970/base/wires/BatteryControlMode.proto:16]: Fixed up

File: proto/zepben/protobuf/cim/extensions/iec61970/base/wires/BatteryControlMode.proto [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+ */
+enum BatteryControlMode {
+
+    /**
+     * Unknown control mode.
+     */
+    UNKNOWN = 0;
+
+    /**
+     * This mode directs the BatteryUnit to discharge as needed to maintain the power level of the monitored element within a defined range
```
[the remark this answers is not in the document]
Comment (Esaotuelr) [reply proto/zepben/protobuf/cim/extensions/iec61970/base/wires/BatteryControlMode.proto:26]: Added to all of them

File: proto/zepben/protobuf/cim/extensions/iec61970/base/wires/BatteryControl.proto [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+
+
+syntax = "proto3";
+option java_multiple_files = true;
+option java_package = "com.zepben.protobuf.cim.extensions.iec61970.wires";
+option csharp_namespace = "Zepben.Protobuf.CIM.Extensions.IEC61970.Wires";
+package zepben.protobuf.cim.extensions.iec61970.wires;
+
+import "zepben/protobuf/cim/iec61970/base/wires/RegulatingControl.proto";
+import "google/protobuf/struct.proto";
```
[the remark this answers is not in the document]
Comment (Esaotuelr) [reply proto/zepben/protobuf/cim/extensions/iec61970/base/wires/BatteryControl.proto:17]: Wasn't needed, left over from copy paste. Removed.

File: proto/zepben/protobuf/nc/nc-data.proto [hunk 10]
```diff
@@ -179,6 +182,10 @@ message NetworkIdentifiedObject {
         zepben.protobuf.cim.iec61970.base.wires.ReactiveCapabilityCurve reactiveCapabilityCurve = 79;
         zepben.protobuf.cim.iec61970.base.wires.GroundingImpedance groundingImpedance = 80;
         zepben.protobuf.cim.iec61970.base.wires.PetersenCoil petersenCoil = 81;
+        zepben.protobuf.cim.iec61970.base.wires.StaticVarCompensator staticVarCompensator = 82;
+        zepben.protobuf.cim.extensions.iec61968.metering.PanDemandResponseFunction PanDemandResponseFunction = 83;
```
[the remark this answers is not in the document]
Comment (Esaotuelr) [reply proto/zepben/protobuf/nc/nc-data.proto:186]: fixed

File: spec/ewb/IEC61970/Base/Generation/Production/BatteryUnit.yaml [hunk 11]
```diff
@@ -11,4 +11,13 @@ attributes:
...
+- PowerElectronicsUnit
+associations:
+- source: BatteryUnit
+  target: BatteryControl
+  sourceCardinality: 0..*
+  sourceName: BatteryUnit
+  sourceDescription: All controls present on this battery.
+  targetCardinality: 0..1
+  targetName: BatteryControl
+  targetDescription: Battery control on this battery unit.
```
[the remark this answers is not in the document]
Comment (Esaotuelr) [reply spec/ewb/IEC61970/Base/Generation/Production/BatteryUnit.yaml:23]: added

File: spec/ewb/IEC61968/Metering/ControlledAppliance.yaml [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+    type: Boolean
+    description: True if the appliance is a smart appliance.
+  - name: isStripAndBaseboardHeater
+    type: Boolean
+    description: True if the appliance is a strip or baseboard heater.
+  - name: isWaterHeater
+    type: Boolean
+    description: True if the appliance is a water heater.
+ancestors:
+- IdentifiedObject
```
[the remark this answers is not in the document]
Comment (Esaotuelr) [reply spec/ewb/IEC61968/Metering/ControlledAppliance.yaml:41]: Changed up. and in the proto too.

File: spec/ewb/IEC61968/Metering/EndDevice.yaml [hunk 13]
```diff
@@ -25,4 +25,12 @@ associations:
...
\ No newline at end of file
+  targetDescription: Service location whose service delivery is measured by this end device.
+- source: EndDevice
+  target: EndDeviceFunction
+  sourceCardinality: 0..*
+  sourceName: EndDevices
+  sourceDescription: All end device functions this end device performs.
+  targetCardinality: 0..*
+  targetName: EndDeviceFunction
+  targetDescription: End device function available on this device
```
[the remark this answers is not in the document]
Comment (Esaotuelr) [reply spec/ewb/IEC61968/Metering/EndDevice.yaml:36]: I think my understanding of the association is backwards. I had a chat with Krisheel about it yesterday (regarding all the associations) but I think I just end up confusing him about it too. wasn't copied from CIM100 version, but implemented backwards on wrong understanding of the association between the two.

[the remark this answers is not in the document]
Comment (Esaotuelr) [reply spec/ewb/IEC61968/Metering/EndDevice.yaml:36]: Changed up to be inline with the README.

File: spec/ewb/IEC61968/Metering/EndDeviceFunction.yaml [hunk 14]
```diff
@@ -0,0 +1,8 @@
+name: EndDeviceFunction
+description: Function performed by an end device such as a meter, communication equipment, controllers, etc.
+attributes:
+  - name: enabled
+    type: Boolean
+    description: True if the function is enabled.
+ancestors:
+- AssetFunction
```
[the remark this answers is not in the document]
Comment (Esaotuelr) [reply spec/ewb/IEC61968/Metering/EndDeviceFunction.yaml:8]: added

File: spec/ewb/IEC61970/Base/Generation/Production/BatteryUnit.yaml (hunk 11 above)
[the remark this answers is not in the document]
Comment (Esaotuelr) [reply spec/ewb/IEC61970/Base/Generation/Production/BatteryUnit.yaml:23]: Fixed according to README

File: spec/ewb/IEC61970/Base/Wires/StaticVarCompensator.yaml (hunk 4 above)
Comment (Esaotuelr) [reply spec/ewb/IEC61970/Base/Wires/StaticVarCompensator.yaml:1]: Changed it up by Copying and modified.

File: proto/zepben/protobuf/cim/iec61970/base/wires/generation/production/BatteryUnit.proto [hunk 15]
```diff
@@ -42,4 +43,9 @@ message BatteryUnit {
    */
   int64 storedE = 4;
 
+  /**
+   * [ZBEX] All controls present on this battery.
+   */
+  repeated extensions.iec61970.wires.BatteryControl batteryControls = 5;
```
[the remark this answers is not in the document]
Comment (Esaotuelr) [reply proto/zepben/protobuf/cim/iec61970/base/wires/generation/production/BatteryUnit.proto:49]: Changed it up as my understanding of the association is backwards.

File: spec/ewb/extensions/IEC61968/Metering/PanDemandResponseFunction.yaml (hunk 5 above)
Comment (Esaotuelr) [reply spec/ewb/extensions/IEC61968/Metering/PanDemandResponseFunction.yaml:2]: Added to the lot, and also added the missing attribute to regulatingcontrol.

File: spec/ewb/extensions/IEC61970/Base/Wires/BatteryControl.yaml (hunk 6 above)
Comment (Esaotuelr) [reply spec/ewb/extensions/IEC61970/Base/Wires/BatteryControl.yaml:5]: Changed it up and went through the other file to link up what I think is correct.

File: proto/zepben/protobuf/cim/iec61968/metering/EndDevice.proto [hunk 16]
```diff
@@ -49,4 +50,9 @@ message EndDevice {
      */
     string serviceLocationMRID = 4;
 
+    /**
+     * All end device functions this end device performs.
+     */
+    repeated EndDeviceFunctionKind endDeviceFunctions = 5;
```
[the remark this answers is not in the document]
Comment (Esaotuelr) [reply proto/zepben/protobuf/cim/iec61968/metering/EndDevice.proto]: changed

**Review by charlta — APPROVED** [review]

**Pull request merged**